withspectrum/spectrum
# Weekly + daily digest email

Hey @mxstbr - 

Here's my first attempt at a weekly digest builder. I've set up a new worker called Chronos:
*(Chronos (/ˈkroʊnɒs/; Greek: Χρόνος, "time", pronounced [kʰrónos], also transliterated as Khronos or Latinised as Chronus) is the personification of Time)*

I'd love your feedback on the implementation details of this - I'm trying to be sensitive to performance, even though this will only run weekly. Eventually we can fork this and have it run daily for people who want daily digests!

**To set up:**
1. `yarn install`
2. `yarn run dev:chronos`
3. We will have to run a migration to update all `usersSettings` records. I didn't feel like doing this, so you can just do:

```
r.db('spectrum')
  .table('usersSettings')
  .filter({ userId: YOUR_LOCAL_USER_ID })
  .update({
    notifications: {
      types: {
        newMessageInThreads: {
         email: r.row('notifications')('types')('newMessageInThreads')(
           'email'
         ),
        },
        newThreadCreated: {
         email: r.row('notifications')('types')('newThreadCreated')(
           'email'
         ),
        },
        newDirectMessage: {
         email: r.row('notifications')('types')('newDirectMessage')(
           'email'
         ),
        },
        weeklyDigest: {
         email: true
        },
      },
    },
  })
```

**To test:**
I set up an api route just to quickly trigger the script; obviously we'll switch this to be triggered from a chron job. For now: ping `http://localhost:3001/api/weekly-digest` - that request doesn't respond with anything, so just stop the request after a second or so in the browser.

You should see something like this:
<img width="943" alt="screenshot 2017-08-31 19 16 09" src="https://user-images.githubusercontent.com/1923260/29952845-e0917b92-8e80-11e7-93bc-5fc47f3b84e3.png">

The script fails at the last section to send an email because I haven't wired that up yet. Should be trivial to do now that I have the data.

Note that the final data will be:

```js
[
  {
    email: String,
    name?: String,
    userId: ID,
    threads: [ { ThreadType }, { ThreadType }, ... ]
  }
]

type ThreadType = {
  community: {
    name: String,
    slug: String,
    profilePhoto: String
  },
  channelId: ID,
  threadId: ID,
  title: String
}
```

### The email:
My goal is to render something like this:
<img width="718" alt="messages" src="https://user-images.githubusercontent.com/1923260/29953141-a38d2852-8e82-11e7-8582-eb4217181e6d.png">

It should be trivial to get extra meta data with the way I've written the script.

### How it works:
**Criteria for a weekly digest**
- [x] the digest must have more than 3 threads in order to be triggered
- [x] each thread in the digest must have at least 10 messages
- [x] a weekly digest should show at most the top 10 threads across all of a user's channels (note: I haven't done this yet, just remembered)

**Steps**
1. Start by fetching *all* threads that were active in the past week across all of Spectrum. This might be a perf issue down the road; for now it's just a few thousand records.
2. For each thread, get the message count
3. Filter out threads with message count under 10
4. For each thread, fetch the community name, slug, and profile photo
5. Sort all threads in a key:value object where the `key` is a channel id and `value` is the thread + community data
6. Get all users who have their notification preferences for weekly digests set to `true`
7. For each of those users, fetch their `usersChannels` `channelIds`
8. For each user, determine which threads found in step 5 were posted within a channelId determined in step 7
9. Combine all the above data into one final object containing the user's email + name + id, and an array of all the threads that meet the weekly digest criteria (sorted desc by message count)
10. For each record in this final array, send an email (todo)

It's hard to describe the steps super clearly - there's a lot of shit going on. What is important to me is that the ordering is loosely right, and that we're being as efficient as we could be at this point in time. My goal was to have the final data output be *exactly* what's needed to render the email.
